### DISCUSSION TITLE
more generalisation?

### DISCUSSION THREAD

**Discussion by steffenschumacher** [tobiasge/netbox-initializers#21] [Ideas]
Seems to me, that there are room for generalisation and more code reuse looking at this code.
I have some ideas for this, so its less work for new/updated objects in netbox.
Specifically I would move more logic into the BaseInitializer class, so the load_data method is generic and then initialisers overload/implement other methods to specialise as needed.
Once I have a draft, I will post it here for review?

**Comment by steffenschumacher** [comment]
I've added a draft of my ideas here: https://github.com/steffenschumacher/netbox-initializers/tree/heavy_base
TL;DR;
I've generalised the ```load_data```method, so that each ```BaseInitializer```-derived class 'only' needs to override specific properties instead of implement this method.
Main benefit is that adding support for new NB models is dead simple, and if netbox dictates generic changes across all models, it can be addressed centrally in the ```BaseInitializer```class vs in each of the nearly 50 derived classes..
As an example I've added support for allowing updates beyond just creating absent objects

One could argue the current implementation is more straight forward, however I'd guess my approach is more maintainable in the long run..

I've gone with overriding properties mostly to make it easy to figure out how to subclass, but simply setting variables directly on the derived classes as initially done with ```data_file_name```would work also. That requires fewer LoC, and is easier to read possibly?

Am I way off track?

**Reply by steffenschumacher** [reply]
@tobiasge I don't know if you haven't seen this, or you're swamped or you just don't like the proposal? 
Either way I would appreciate even negative feedback, since I will need to decide on whether to maintain my own fork with update support or whether I can somehow use your official one? (latter preferred ofc)
I haven't done a PR, since I think that's premature at this stage

**Reply by tobiasge** [reply]
I think it's a  good idea to have a more general approach where it will be easier to implement new models and fields from Netbox. But this has too be a reasonable trade off to the introduced complexity.
From a quick look at the code I noticed that the changes to the registry won't work because you changed names of some initializers so that they don't match the ones used in the `INITIALIZER_ORDER` set.

**Reply by steffenschumacher** [reply]
Ahh right, so in devices.py, the name should be "devices" vs "device"? 
Since the order dict uses the plural form, all the objects can be indexed correctly by adding an 's' to the name when registering them.
```python
def register_initializer(initializer: BaseInitializer):
    if InitializerConfig.is_updatable(initializer.nb_model.__name__):
        initializer.set_updates_allowed()
    INITIALIZER_REGISTRY[initializer.name+'s'] = initializer
```

I've only done devices.py and cables.py - this is more a draft/sketch to consider before moving on..
But the question is more if the introduced complexity is worth it (for others than me who need the update stuff anyways..)
I'm heavily biased, so while I hope so, its for others to judge :)

**Reply by steffenschumacher** [reply]
@tobiasge I guess your silence indicates lack of interest in this or?
I think we'll start a fork in a couple of weeks or so, since we are currently unable to upgrade without these features, and we'll probably also need to customise bits and pieces anyways..
Thanks..

**Reply by monsdar** [reply]
@steffenschumacher has there any activity on your side? Did you fork the plugin and started further development?

**Reply by steffenschumacher** [reply]
Hi,
Not lately, but I have an upcoming slot in a months time, where I should be able to work on this🤞
